components/gitpod-protocol/src/generate-async-generator.ts

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+type Sink<T> = {
+    next: (value: T | void) => void;
+};
+
+/**
+ * Generates an asynchronous generator that yields values based on the provided setup function.
+ *
+ * the setup function that takes a sink and returns a cleanup function.
+ * setup sink object has a `next` method that accepts a value to be pushed to the generator.
+ */
+export async function* generateAsyncGenerator<T>(
```

Comment:
it would be good to write unit tests for it, especially to ensure that we dispose eventually